### PR TITLE
SYM-167 : Fix for Empty Http Request Headers When the Value for the Header Is Null

### PR DESCRIPTION
**Pull request by mvu02**
Fix for Issue: https://github.com/symphonyoss/RestApiClient/issues/46

We were sending http requests to Symphony with headers having an empty value and this was causing an exception to occur. The current solution while waiting for NSwag to accept our fix into their project was to fork the NSwag project and updated the code to include a check for null values before attempting to add the headers to the http request. We will need to update the README and generate.bat file after NSwag has accepted and released our suggested changes.

### CODE REVIEW

**Comment by mvu02** [comment]
Hi @malaysf, @tchitnis could you take a look at the changes?

Thanks,
Minh

**Comment by mvu02** [comment]
The ConvertToString changes are coming from: [this commit](https://github.com/RSuter/NSwag/commit/9155d1ccbbdbb3aef9dc23ebb9a9901b517f1b3a)

**Comment by mvu02** [comment]
The System.Net.Http.Headers.MediaTypeHeaderValue.Parse("application/json"); changes are coming from [this commit](https://github.com/RSuter/NSwag/commit/53355438f2b4571e10936f7f73ecd1c297406c2a#diff-57ab6b542c5556e3857fd227ae623f91)

File: src/SymphonyOSS.RestApiClient/Generated/OpenApi/AgentApi/AgentApi.cs [hunk 1]
```diff
@@ -6543,7 +6733,7 @@ public override object ReadJson(Newtonsoft.Json.JsonReader reader, System.Type o
     
         private System.Type GetObjectSubtype(System.Type objectType, string discriminator)
         {
-            foreach (var type in System.Reflection.CustomAttributeExtensions.GetCustomAttributes<JsonInheritanceAttribute>(System.Reflection.IntrospectionExtensions.GetTypeInfo(objectType), false))
+            foreach (var type in System.Reflection.CustomAttributeExtensions.GetCustomAttributes<JsonInheritanceAttribute>(System.Reflection.IntrospectionExtensions.GetTypeInfo(objectType), true))
```
Comment (tchitnis) [inline src/SymphonyOSS.RestApiClient/Generated/OpenApi/AgentApi/AgentApi.cs]: We should verify where this change is coming from as well since we seem to be flipping the flag here.

File: src/SymphonyOSS.RestApiClient/Generated/OpenApi/generate.bat [hunk 2]
```diff
@@ -37,15 +37,13 @@ dotnet ..\..\..\..\tools\SwaggerOperationIdGenerator\SwaggerOperationIdGenerator
...
-
-
 rmdir /s /q AgentApi
 rmdir /s /q AuthenticatorApi
 rmdir /s /q PodApi
 
-call nswag run Config\agentAPI.nswag
-call nswag run Config\authenticatorAPI.nswag
-call nswag run Config\podAPI.nswag
+call ..\..\..\..\..\NSwagBinary\net461\NSwag.exe run Config\agentAPI.nswag
```
Comment (tchitnis) [inline src/SymphonyOSS.RestApiClient/Generated/OpenApi/generate.bat]: We should revert these changes if the NSwag PR is accepted before we merge this one

File: README.md [hunk 3]
```diff
@@ -68,10 +68,12 @@ The library depends on code that has been generated from Symphony's YAML spec fi
 
 In case there is a need to regenerate the code from Symphony's YAML specs:
 
- 1. Make sure Node Package Manager is installed
- 2. Install nswag via command line:  `npm install nswag -g`
- 3. Install api-spec-converter via command line: `npm install api-spec-converter -g`
- 4. Run SymphonyOSS.RestApiClient\Generated\OpenApi\generate.bat
+ 1. Clone [this](https://github.com/mvu02/NSwag) repo to get a local copy of the NSwag code
```
Comment (tchitnis) [inline README.md]: We should revert these changes if the NSwag PR is accepted before we merge this one

Comment (mvu02) [reply README.md]: We could scrap this whole PR if NSwag PR is accepted before we merge this one and jstu do a clean rebuild using nswag npm.

File: src/SymphonyOSS.RestApiClient/Generated/OpenApi/AgentApi/AgentApi.cs (hunk 1 above)
Comment (mvu02) [reply src/SymphonyOSS.RestApiClient/Generated/OpenApi/AgentApi/AgentApi.cs]: https://github.com/mvu02/NJsonSchema/commit/0c733d2681005615fda7cb745d52b1cf61ac3d36

**Comment by mvu02** [comment]
From the RPD, Johan wants these additional functionalities which I have not yet added:
```
In addition to the above, we should also build the functionality to:

1. Respond in chatrooms and mims only when mentioned

2. Send "no content" as a response if needed

3. Respond with help test in a im but with no response in mims and chat rooms for general messages.
```

**Comment by malaysf** [comment]
@mvu02 was that message meant for another PR?

**Comment by malaysf** [comment]
@mvu02 Now that your nswag changes have been accepted, can you update the instructions to not require your personal fork? At this point could we instead just reference a minimum version of nswag?

**Comment by mvu02** [comment]
@malaysf Woops, yes that is meant for another PR, I will remove that comment. And it doesn't look like RSuter has released the next version that includes our changes yet, but I will update the instructions to not use my personal fork. Hopefully RSuter will release the next version soon so I can directly reference that minimum version.
